File: R/header_to_fwf_position.R
#' Converts header line into a vector of column names
#'
#' @keywords internal
#'
#' @param header A length-one character vector of a header line
#' from a DSSAT format output file
#'
#' @param left_justified A character vector of column names that
#' should be left justified
#'
#' @return a tibble containing the begin position, end position,
#'  and name for each column from the original header line
#'
#' @importFrom stringr str_replace_all str_detect str_c str_split str_subset
#' @importFrom purrr map_lgl map_chr
#' @importFrom dplyr "%>%" mutate arrange tibble filter last pull
#' @importFrom readr fwf_positions
#'
#' @examples
#'
#' header_line <-
#'   "@YEAR DOY   DAS   SRAA    ES1D    ES2D    ES3D    ES4D    ES5D    ES6D    ES7D    ES8D"
#' DSSAT:::header_to_fwf_position(header_line)

header_to_fwf_position <- function(header, left_justified='EXCODE',
                                   col_types=NULL, col_names=NULL,
                                   read_only = NULL){

  # Replace @ sign and | with space
  header <- header %>%
    str_replace_all(c('@'=' ',
                      '\\|'=' ',
                      '!.*'=''))

  if(is.null(read_only)){
    # Combine left justified,col_types and col_names for known columns
    col_names <- c(col_names,left_justified,names(col_types$cols)) %>%
      unique()
    col_names <- col_names %>%
      name_to_regex() %>%
      map_lgl(~str_detect(header,.)) %>%
      {col_names[.]}
  }else{
    col_names <- read_only %>%
      str_c('((^)|( ))',.,'(( )|($))') %>%
      map_lgl(~str_detect(header,.)) %>%
      {read_only[.]}
  }

  if(is.null(read_only)){
    # reduce to unknown column names
    if(length(col_names) >= 1){
      unknown <- col_names %>%
        name_to_regex() %>%
        str_c('(',.,')') %>%
        str_c(collapse='|') %>%
        str_remove_all(header,.)
    }else{
      unknown <- header
    }
    # Extract unknown column names and combine with known col_names
    cnames <- unknown %>%
      str_split('(?<=[^ ]) +(?=[^ ])') %>%
      unlist() %>%
      str_subset('^ *$',negate = TRUE) %>%
      str_remove_all('(^ +)|( +$)') %>%
      c(col_names,.)
  }else{
    cnames <- col_names
  }

  if(is.null(read_only)){
    # Infer column positions based on header
      cnames_regex <- cnames %>%
      map_chr(function(name){
        if(any(str_detect(name,name_to_regex(left_justified)))){
          regex <- name %>%
            name_to_regex() %>%
            str_c(' *')
        }else if(!any(str_detect(de_regex(name),name_to_regex(col_names)))){
          regex <- name %>%
            name_to_regex() %>%
            str_c(' *',.,'((?= )|(?=$))')
        }else{
          regex <- name %>%
            name_to_regex()
        }
        return(regex)
      })
  }else{
    cnames_regex <- str_c('\\.*',col_names,'\\.*')
  }

  start_end <- get_start_end(header,cnames_regex) %>%
    {tibble(start = as.numeric(.[,1]),
            end = as.numeric(.[,2]))} %>%
    mutate(col_names = {cnames %>%
                        de_regex() %>%
                        str_remove_all('(^ *)|( *$)')},
           cnames = cnames,
           regex = cnames_regex) %>%
    arrange(.data$start)

  # Set start of first column to 1
  start_end$start[1] <- 1

  # Reconcile gaps/overlaps in column bounds
  start_end <- reconcile_gaps(start_end,left_justified)

  # Check column positions for left justified cases
  last_left_justified <- start_end %>%
    pull(cnames) %>%
    last() %>%
    {. %in% left_justified}
  # Set end of last column to NA (i.e. unbounded)
  if(last_left_justified)  start_end$end[nrow(start_end)] <- NA

  # Convert column positions to fixed widths for use with read_fwf()
  fwf_pos <- start_end %>%
    {fwf_positions(start=.$start,end=.$end,col_names=.$col_names)} %>%
    filter(col_names!='-99')

  if(!is.null(read_only)){
    fwf_pos <- fwf_pos %>%
      filter(col_names %in% read_only)
  }

  return(fwf_pos)
}

File: src/header_to_fwf_position.cpp
#include <cpp11.hpp>
#include <string>
#include <regex>
#include <numeric>
#include <iostream>
using namespace cpp11;

void reconcile_gaps_cpp(writable::strings &cnames,
                        writable::strings &regexpr,
                        writable::strings &col_names,
                        writable::doubles &start,
                        writable::doubles &end,
                        writable::strings &left_justified);

void get_start_end(std::string &header,
                   strings &regex,
                   writable::doubles &start,
                   writable::doubles &end);

cpp11::writable::strings de_regex(strings the_regex);

cpp11::writable::strings gen_regex(strings cnames,
                                   strings ljust);

void arrange_by_start(cpp11::writable::doubles &beg_in,
                      cpp11::writable::doubles &end_in,
                      cpp11::writable::strings &regex_in,
                      cpp11::writable::strings &col_names,
                      cpp11::writable::strings &cnames);

cpp11::writable::strings extract_names(std::string header,
                                       strings col_types,
                                       strings ljust,
                                       strings col_names);

[[cpp11::register]]
data_frame header_to_fwf_position_cpp(std::string header,
                                      strings col_types_names,
                                      strings left_justified,
                                      strings col_names_in,
                                      strings read_only) {
  // Declare incoming variables
  writable::strings cnames_in;
  writable::strings regex_in;
  writable::strings names_in;

  cnames_in = extract_names(header, col_types_names, left_justified,
                            col_names_in);

  regex_in = gen_regex(cnames_in, left_justified);

  names_in = de_regex(cnames_in);

  // Declare temporary variables
  writable::strings ljust;
  writable::doubles beg_in;
  writable::doubles end_in;

  // Declare output variables
  writable::doubles begin;
  writable::doubles end;
  writable::strings cnames;
  writable::strings col_names;

  // Remove leading/trailing spaces from col_names
  for(auto it = names_in.begin(); it != names_in.end(); ++it){
    *it = std::regex_replace(std::string(r_string(*it)),
                             std::regex("(^ *)|( *$)"),
                             "");
  }

  // Extract cnames from header
  for(auto it = regex_in.begin(); it != regex_in.end(); ++it){
    std::regex tmp_regex = std::regex(
      std::regex_replace(std::string(r_string(*it)),
                       std::regex(" [*+]"),
                       "")
    );
    std::smatch match;
    std::regex_search(header, match, tmp_regex);
    cnames.push_back(match.str());
  }

  // Infer column positions from header
  get_start_end(header, regex_in, beg_in, end_in);

  // Arrange by start column
  arrange_by_start(beg_in, end_in, regex_in, names_in, cnames);
  // beg_in, end_in, regex_in, names_in, cnames

  // Set start of first column to 1
  beg_in[0] = 1;

  // Remove extra spaces involved in regular expressions from left_justified
  if(left_justified.size() > 0){
    for(auto it = left_justified.begin(); it != left_justified.end(); ++it){
      std::string check = std::regex_replace(std::string(*it), std::regex("( \\*)|( \\+)"), "");
      std::smatch the_match;
      if(std::regex_search(header, the_match, std::regex(check))) ljust.push_back(the_match.str());
    }
  }

  // Reconcile gaps/overlaps in column bounds
  reconcile_gaps_cpp(cnames, regex_in, names_in, beg_in, end_in, ljust);

  // Set values for begin, end, and col_names
  for(size_t i = 0; i < beg_in.size(); ++i){
    // Check against read_only
    bool store = read_only.size() <= 0 && r_string(names_in[i]) != strings({"-99"})[0];
    if(!store){
      for(auto it = read_only.begin(); it != read_only.end(); ++it){
        store = names_in[i] == *it;
        if(store) break;
      }
    }
    if(store){
      // Set begin indices minus one
      begin.push_back(beg_in[i] - 1.);
      end.push_back(end_in[i]);
      col_names.push_back(names_in[i]);
      if(i == beg_in.size() - 1){
        // Check column positions for left justified cases
        for(auto it = ljust.begin(); it != ljust.end(); ++it){
          // Set end of last column to NA (i.e. unbounded)
          if(cnames[i] == *it){
            end[end.size() - 1] = NA_REAL;
            break;
          }
        }
      }
    }
  }

  // Construct results data.frame
  writable::data_frame results({
    "begin"_nm = begin,
    "end"_nm = end,
    "col_names"_nm = col_names
  });

  // Set results class to tibble
  results.attr("class") = strings({"tbl_df", "tbl", "data.frame"});

  return results;
}
